src/Microsoft.AspNetCore.Blazor.Browser/Interop/ArgumentList.cs

```diff
@@ -0,0 +1,270 @@
+                    default:
+                        throw new InvalidOperationException($"Unsupported number of arguments '{arguments.Length}'");
+                }
+            }
+        }
+
+        public static Func<string, ArgumentList> GetDeserializer(Type deserializedType)
+        {
+            if (_deserializers.TryGetValue(deserializedType, out var deserializer))
+            {
+                return deserializer;
+            }
+
+            switch (deserializedType.GetGenericArguments().Length)
+            {
+                case 0:
+                    deserializer = JsonDeserialize;
+                    break;
+                case 1:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize1", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 2:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize2", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 3:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize3", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 4:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize4", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 5:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize5", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 6:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize6", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 7:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize7", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                case 8:
+                    deserializer = (Func<string, ArgumentList>)deserializedType.GetMethod("JsonDeserialize8", DeserializeFlags)
+                        .CreateDelegate(typeof(Func<string, ArgumentList>));
+                    break;
+                default:
+                    throw new InvalidOperationException("Shouldn't have gotten here!");
+            }
+
+            _deserializers[deserializedType] = deserializer;
```

Comment:
This code isn't thread-safe. This will work for now with the browser, but in the general case it's kinda busto. You could consider exposing the deserializer as a static property on the type so that .NET just does the caching for you.